qubes/backup.py

```diff
@@ -630,39 +630,44 @@ async def _wrap_and_send_files(self, files_to_backup, output_queue):
-                    # also use our tar writer when renaming file
-                    assert not stat.S_ISDIR(file_stat.st_mode), \
-                        "Renaming directories not supported"
-                    tar_cmdline = ['python3', '-m', 'qubes.tarwriter',
-                        '--override-name=%s' % (
-                            os.path.join(file_info.subdir, os.path.basename(
-                                file_info.name))),
-                        path]
-                if self.compressed:
-                    tar_cmdline.insert(-2,
-                        "--use-compress-program=%s" % self.compression_filter)
-
-                self.log.debug(" ".join(tar_cmdline))
-
-                # Pipe: tar-sparse | scrypt | tar | backup_target
-                # TODO: log handle stderr
-                # pylint: disable=not-an-iterable
-                tar_sparse = await asyncio.create_subprocess_exec(
-                    *tar_cmdline, stdout=subprocess.PIPE)
+
+                fd = os.open(path, os.O_RDONLY|o.O_NOCTTY|os.O_CLOEXEC)
+                try:
+                    fd_path = '/proc/self/fd/' + str(fd)
+                    tar_cmdline = (["tar", "-Pc", '--sparse',
+                                    '-C', os.path.dirname(path)] +
+                                   (['--dereference'] if
+                                   file_info.subdir != "dom0-home/" else []) +
+                                   ['--xform=s:^%s:%s\\0:' % (
+                                       os.path.basename(path),
+                                       file_info.subdir),
+                                       fd_path,
+                                   ])
+                    file_stat = os.stat(path)
+                    if stat.S_ISBLK(file_stat.st_mode) or \
+                            file_info.name != os.path.basename(path):
+                        # tar doesn't handle content of block device, use our
+                        # writer
+                        # also use our tar writer when renaming file
+                        assert not stat.S_ISDIR(file_stat.st_mode), \
+                            "Renaming directories not supported"
+                        tar_cmdline = ['python3', '-m', 'qubes.tarwriter',
+                            '--override-name=%s' % (
+                                os.path.join(file_info.subdir, os.path.basename(
+                                    file_info.name))),
+                            fd_path]
+                    if self.compressed:
+                        tar_cmdline.insert(-2,
+                            "--use-compress-program=" + self.compression_filter)
+
+                    self.log.debug(" ".join(tar_cmdline))
+
+                    # Pipe: tar-sparse | scrypt | tar | backup_target
+                    # TODO: log handle stderr
+                    tar_sparse = await asyncio.create_subprocess_exec(
+                        *tar_cmdline, stdout=subprocess.PIPE, keep_fds=(fd,))
```

Comment:
Can you tell _exactly_ what's the point in opening FD in qubesd, when it's taken over by a process started few lines below? If qubesd is terminated before this process starts - nothing would try to open the exported path (so it can be cleaned up). If its terminated after (but before backup finishes), then the backup process will be interrupted anyway (see the stdout of this process is handled in qubesd - so terminating qubesd _will_ interrupt the backup. It doesn't matter if tar will manage to open that path or not.

Reply:
Good point.  I guess this isn’t strictly necessary in that case.  That said, it does make the code easier to reason about.  With this approach, it is _obvious_ that there are no races, whereas this is not the case with the current approach.  It also avoids having to figure out proper escaping for the match part of `--xform`.